### DISCUSSION TITLE
My update script doesn't work any more as expected

### DISCUSSION THREAD

**Discussion by curiosityseeker** [netblue30/firejail#5314] [Q&A] [answered]
I've been using [this](https://stackoverflow.com/questions/3258243/check-if-pull-needed-in-git/3278427#3278427) approach for a long time to make sure that git pull is only executed if the Firejail remote repository has changed:

```
#!/bin/sh
cd ~/firejail

git remote update

UPSTREAM=${1:-'@{u}'}
LOCAL=$(git rev-parse @)
REMOTE=$(git rev-parse "$UPSTREAM")
BASE=$(git merge-base @ "$UPSTREAM")

if [ $LOCAL = $REMOTE ]; then
    echo -e "\n   Firejail is up-to-date!"
    echo
elif [ $LOCAL = $BASE ]; then

git pull --no-rebase
./configure --prefix=/usr --enable-apparmor
make
sudo /home/seeker/bin/firejailinstall

fi
```
However, since a couple of weeks it doesn't work any more - git pull is **always** executed. It seems that github changed something - but what? Any ideas?

**Comment by ghost** [comment]
> However, since a couple of weeks it doesn't work any more - git pull is always executed. It seems that github changed something - but what? Any ideas?

Are you still having these problems? I'm not actively tracking GitHub changes so can't offer anything useful here, but it _might_ have been a temporary thing, happens quite frequently on GH IMO...

**Comment by curiosityseeker** [comment, marked as answer]
Oops - sorry, I had forgotten about this issues! Thanks for reminding me!

I've found an easy workaround that works reliably for me so far:

```
git remote update
FILE="/tmp/mydiff.txt"
git status 2>&1 | tee "$FILE"
STRING="Commits hinter 'origin/master'" || "Commit hinter 'origin/master'"
if grep -q "$STRING" "$FILE"; then
    git pull --no-rebase
    ./configure --prefix=/usr --enable-apparmor
    make
    sudo /home/seeker/bin/firejailinstall
    echo -e "\n   Firejail updated successfully";
else
    echo -e "\n   Firejail is up-to-date!"
fi
```
Needless to say that STRING must be adjusted according to your language.

**Comment by rusty-snake** [comment]
OT: Why do you habe a firejailinstall  script instead of a plain `make install`? Does it save a list of installed files?

**Reply by curiosityseeker** [reply]
> OT: Why do you habe a firejailinstall script instead of a plain `make install`? Does it save a list of installed files?

Good question ;) Because in that file I've included some more commands similar to the examples provided [here](https://github.com/netblue30/firejail/wiki/Using-firejail-from-git#Makefile). And it's one of the few commands added to my /etc/sudoers that can be executed password-free to make updating my system a bit easier.

**Reply by rusty-snake** [reply]
> And it's one of the few commands added to my /etc/sudoers that can be executed password-free to make updating my system a bit easier.

To make becoming root easier ;)

I can just replace it with a malicious one and execute it without interaction.

**Reply by curiosityseeker** [reply]
> > And it's one of the few commands added to my /etc/sudoers that can be executed password-free to make updating my system a bit easier.
> 
> To make becoming root easier ;)
> 
> I can just replace it with a malicious one and execute it without interaction.

Hm, how would you remotely replace that file? I'm genuinely interested!

FWIW, this is it:
```
 #!/bin/sh
 
 cd /home/seeker/firejail
 make install

 sed -i 's/# force-nonewprivs no/force-nonewprivs yes/' /etc/firejail/firejail.config
 sed -i 's/# bind yes/bind no/' /etc/firejail/firejail.config
 sed -i 's/# cgroup yes/cgroup no/' /etc/firejail/firejail.config
 sed -i 's/# overlayfs yes/overlayfs no/' /etc/firejail/firejail.config

 groupadd firejail
 chown -c root:firejail /usr/bin/firejail
 chmod -c 4750 /usr/bin/firejail
 usermod -a -G firejail $USER

 
 # rm /usr/local/bin/brave
 # rm /usr/local/bin/brave-browser-nightly
 rm /usr/local/bin/claws-mail
 rm /usr/local/bin/flameshot
 rm /usr/local/bin/gwenview
 rm /usr/local/bin/keepass
 rm /usr/local/bin/keepassxc
 # rm /usr/local/bin/okular
 rm /usr/local/bin/signal-desktop
 # rm /usr/local/bin/firefox
 rm /usr/local/bin/thunderbird
 rm /usr/local/bin/cherrytree
 rm /usr/local/bin/spectacle
 rm /usr/local/bin/VirtualBox
 rm /usr/local/bin/zim
 
 sed -i 's/flameshot/#flameshot/' /etc/firejail/firecfg.config
 sed -i 's/claws-mail/#claws-mail/' /etc/firejail/firecfg.config
 sed -i 's/gwenview/#gwenview/' /etc/firejail/firecfg.config
 sed -i 's/keepass/#keepass/' /etc/firejail/firecfg.config
 sed -i 's/keepassxc/#keepassxc/' /etc/firejail/firecfg.config
 # sed -i 's/okular/#okular/' /etc/firejail/firecfg.config
 sed -i 's/signal-desktop/#signal-desktop/' /etc/firejail/firecfg.config
 sed -i 's/thunderbird/#thunderbird/' /etc/firejail/firecfg.config
 sed -i 's/cherrytree/#cherrytree/' /etc/firejail/firecfg.config
 sed -i 's/spectacle/#spectacle/' /etc/firejail/firecfg.config
 sed -i 's/VirtualBox/#VirtualBox/' /etc/firejail/firecfg.config
 sed -i 's/zim/#zim/' /etc/firejail/firecfg.config
 
 firecfg
```

**Reply by rusty-snake** [reply]
I mean as an attacker who has user privileges.

**Reply by rusty-snake** [reply]
As an idea to simplify maintenance: 
```bash
sed -i \
        -e 's/# force-nonewprivs no/force-nonewprivs yes/' \
        -e 's/# bind yes/bind no/' \
        -e 's/# cgroup yes/cgroup no/' \
        -e 's/# overlayfs yes/overlayfs no/' \
        /etc/firejail/firejail.config

# Only needed once after new installation of the system: groupadd firejail
chown -c root:firejail /usr/bin/firejail 
chmod -c 4750 /usr/bin/firejail 
usermod -a -G firejail $USER 

FIRECFG_IGNORE=( 
        claws-mail
        flameshot
        gwenview
        keepass
        keepassxc
        signal-desktop
        thunderbird
        cherrytree
        spectacle
        VirtualBox
        zim
)
 
for app in "${FIRECFG_IGNORE[@]}"; do
        rm /usr/local/bin/"$app"
        sed -i "s/$app/#$app/" /etc/firejail/firecfg.config           
done    
 
firecfg
```

**Reply by curiosityseeker** [reply]
> I mean as an attacker who has user privileges.

~/bin is blacklisted in my globals.local and in my AppArmor profiles. So I think that risk is manageable ;-)

**Reply by curiosityseeker** [reply]
> As an idea to simplify maintenance:
> 
> ```shell
> sed -i \
>         -e 's/# force-nonewprivs no/force-nonewprivs yes/' \
>         -e 's/# bind yes/bind no/' \
>         -e 's/# cgroup yes/cgroup no/' \
>         -e 's/# overlayfs yes/overlayfs no/' \
>         /etc/firejail/firejail.config
> 
> # Only needed once after new installation of the system: groupadd firejail
> chown -c root:firejail /usr/bin/firejail 
> chmod -c 4750 /usr/bin/firejail 
> usermod -a -G firejail $USER 
> 
> FIRECFG_IGNORE=( 
>         claws-mail
>         flameshot
>         gwenview
>         keepass
>         keepassxc
>         signal-desktop
>         thunderbird
>         cherrytree
>         spectacle
>         VirtualBox
>         zim
> )
>  
> for app in "${FIRECFG_IGNORE[@]}"; do
>         rm /usr/local/bin/"$app"
>         sed -i "s/$app/#$app/" /etc/firejail/firecfg.config           
> done    
>  
> firecfg
> ```

Nice! Thanks a lot for that! My original list was much smaller, and when it grew larger I didn't bother to make any changes. But your solution looks really good, indeed!
